pkgs/build-support/setup-hooks/sphinx.sh

```diff
@@ -0,0 +1,93 @@
+## Sphinx documentation generator can generate documentation in multiple
+## formats, and different users prefer different formats.
+##
+## This hook takes care of building and installing documentation in
+## multiple formats into separate outputs, so user can pick only what he
+## wants.
+##
+## Since "outputs" argument to "mkDerivation" is fundamentally
+## hard-wired into Nix, derivation author is still responsible in value
+## of "outputs" argument.
+##
+## This hook only builds documentation if formats that have matching
+## output. E.g, if "pdf" output does not exist, pdf documentation is not
+## built.
+
+# Mapping from output names into Sphinx format name. Most are used
+# verbatim, but "pdf" output is mapped to latexpdf, since latex is
+# implementation detail end user does not care about.
+_resolveSphinxFormat() {
+    case "$1" in
+    (html) echo html ;;
+    (singlehtml) echo singlehtml ;;
+    (man) echo man ;;
+    (info) echo info ;;
+    # Wish there was singletext output
+    (text) echo text ;;
+    (qthelp) echo qthelp ;;
+    (pdf) echo latexpdf ;;
+    esac
+}
+
+buildSphinxPhase() {
+    local format o
+    runHook preBuildSphinx
+
+    # epub is zip, and zip does not support dates before 1980.
+    for o in $outputs ; do
+        format=$(_resolveSphinxFormat "${o}")
+        [[ -n "${format}" ]] || continue
+
+        find . -exec touch -h -d '1990-01-02' '{}' '+'
+        mkdir -p ".sphinx/${o}"
+        sphinx-build -M "${format}" "${sphinxRoot:-doc/src}" ".sphinx/${o}" -v
+    done
+
+    runHook postBuildSphinx
+}
+
+installSphinxPhase() {
+    local format o f
+    runHook preInstallSphinx
+
+    for o in $outputs ; do
+        format=$(_resolveSphinxFormat "${o}")
+        [[ -n "${format}" ]] || continue
+
+        case "$o" in
+        (pdf)
```

Comment:
How much do we care about pdf if we can have html?

Reply:
I don't know. I personally only care about singlehtml. Well, if one day sphinx finally get singletext output, I will stop caring about singlehtml either. It proves nothing, though.